### PR TITLE
[flutter_tools]Add to app supports multiple host apps

### PR DESCRIPTION
**Pull request by kassadin**
## Description

#52791 make host name configable, but not support mulple host.

This pr let _Add to App_  support multiple host apps and any host name, no need to configure `flutter.hostAppProjectName`.

Gradle `FlutterPlugin` will check Android AppPlugin and `implementation project(':flutter')`.

## Related Issues

#42214 

## Tests

Update integration test module_custom_host_app_name_test
> SampleApp and SampleApp2 are flutter host. SampleAppWithout is normal app. They will all work fine

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

No

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jmagman** [comment]
Related https://github.com/flutter/flutter/pull/52791.

**Comment by blasten** [comment]
`flutter.hostAppProjectName` is configured in the host app. Is there an issue to setting that value in the host app?

**Comment by xster** [comment]
@kassadin could you help us understand a bit the issue you're resolving? https://github.com/flutter/flutter/issues/42214 is a bit hard to parse since the original post from that bug is already solved as you pointed out. Could you point to the specific comment that illustrates the issue or open a new issue with reproduction steps?

**Comment by kassadin** [comment]
@xster The android project may have more than one `com.android.application` (host app) which needs to include flutter, like the structure in the test project, https://github.com/flutter/flutter/tree/107b6cf4a3aa3ace85b712ba03a244557070c8ca/dev/integration_tests /android_custom_host_app.

Flutter is only the common ui part of SampleApp and SampleApp2.

The original implementation only supported one host app.

Did I make it clear?

**Comment by zanderso** [comment]
@xster @blasten ping

**Comment by blasten** [comment]
@kassadin Sure, but why not structuring the files such that each host app has its own `gradle.property` file?  This way you can set `flutter.hostAppProjectName`: https://github.com/flutter/flutter/blob/3744b82d8cc3af1e089e45c3435114c4c7356007/dev/integration_tests/android_custom_host_app/gradle.properties#L4

Would that be sufficient to handle the use case?

**Comment by kassadin** [comment]
@blasten Of course, The difference is whether developers are required to configure it. 

We can also make `flutter.hostAppProjectName` support multiple options.like `flutter.hostAppProjectName=app1,app2`

The pr code does not need configuration. 

Do you prefer manual configuration to automatic configuration?

**Comment by blasten** [comment]
I think the current workaround looks reasonable given how common this appears to be. @xster do you agree?

**Comment by zanderso** [comment]
@blasten are you able to take a look at this? It might be some time before @xster is available.

**Comment by blasten** [comment]
yeah. I will close this PR, but feel free to articulate the problem statement and workaround in #52791.  I also wonder if we could just close that issue by updating the docs with the suggested workaround. @kassadin would you be able to help there?
